Add NZBGet as download client

jonjon1123 opened the pull request:
*Thank you for contributing to Homarr! So that your Pull Request can be handled effectively, please populate the following fields (delete sections that are not applicable)*

### Category
> One of: Bugfix / Feature / Code style update / Refactoring Only / Build related changes /  Documentation / Other (Please specify!)

Feature

### Overview
> Briefly outline your new changes...

- Added NZBGet to the Sabnzbd module.
- Changed Sabnzbd naming to "Usenet" to reflect that both NZBGet & Sabnzbd can be used now.
- Implemented NZBGet API

### Issue Number _(if applicable)_
> Related issue: #00

#282

### New Vars _(if applicable)_
> If you've added any new build scripts, environmental variables, config file options, dependency please outline here.

I added a new library that helps in working with the NZBGet api: `"nzbget-api": "0.0.3"`

### Screenshot _(if applicable)_
> If you've introduced any significant UI changes, please include a screenshot.

<img width="1510" alt="image" src="https://user-images.githubusercontent.com/691969/198859216-f239cc5f-3529-47dd-a7f4-fdcc62c8c501.png">
<img width="1510" alt="image" src="https://user-images.githubusercontent.com/691969/198859243-54dd4244-c4a1-4fea-b4ca-838a06d629e4.png">

manuel-rw commented:
Thank you for your awesome contribution to Homarr ❤️🚀

manuel-rw commented on public/locales/es/modules/usenet.json:
```diff
@@ -1,7 +1,7 @@
 {
   "descriptor": {
-    "name": "Sabnzbd",
-    "description": "Te permite ver tu cola e historial de Sabnzbd, pausar y resumir descargas"
```
I don't know what languages you're speaking, but don't **need** to translate, because we have an awesome translation community that does this. But since you already did this, we can just leave it like that.

More info: https://homarr.dev/docs/community/translations

jonjon1123 replied:
Oh ok! Good to know. I just used Google Translate. :)

manuel-rw commented on src/pages/api/modules/usenet/history.ts:
```diff
@@ -33,23 +34,72 @@ async function Get(req: NextApiRequest, res: NextApiResponse) {
+          throw new Error('Error while getting NZBGet history');
+        }
+
+        const nzbgetItems: UsenetHistoryItem[] = nzbgetHistory.map((item: any) => ({
```
Replace type ``any`` with a type definition

jonjon1123 replied:
I am working on getting types created for this (and other return types from the nzbget api)

manuel-rw commented on data/configs/default.json:
```diff
@@ -7,14 +7,40 @@
+      "icon": "https://cdn.jsdelivr.net/gh/walkxcode/dashboard-icons/png/nzbget.png",
+      "url": "http://192.168.1.61:6789",
+      "username": "jonjon1123",
+      "password": "aRsBPQfVxN3u@@5",
```
Ouch...

manuel-rw replied:
![](https://media.tenor.com/vCOInrRvG5UAAAAS/face-palm-oh-no.gif)

jonjon1123 replied:
Yep :( Def accidentally committed the config file... Password has been changes so all is good.

manuel-rw replied:
I don't know why you resolved this. The secret is still here and must be removed!

manuel-rw commented on src/pages/api/modules/usenet/queue.ts:
```diff
@@ -75,6 +75,7 @@ async function Get(req: NextApiRequest, res: NextApiResponse) {
           throw new Error('Error while getting NZBGet status');
         }
 
+        // TODO: Add type for item here
```
Blocker: Resolve TODO

manuel-rw commented on src/pages/api/modules/usenet/nzbget/types.ts:
```diff
@@ -0,0 +1,27 @@
+  DownloadedSizeHi: number;
+  DownloadedSizeMB: number;
+  DownloadTimeSec: number;
+  // TODO: Finish adding the rest of the properties
```
Blocker: Resolve TODO

manuel-rw commented on src/pages/api/modules/usenet/nzbget/types.ts:
```diff
@@ -0,0 +1,27 @@
+  Category: string;
+  FileSizeLo: number;
+  FileSizeHi: number;
+  FileSizeMB: number;
```
Another question: Are the fields of the SABnzbd and NZBget similar? If they have overlapping fields, we should extract them to a base type.

jonjon1123 replied:
From what I can tell, they are not really the same. The nzbget api returns more raw data and the caller has to do the logic to derive meaning (such as percent complete). The sabnzbd api seems to return more of this data already calculated.

JannesV commented:
Looks great! However my only concern is the usage of the `nzbget-api` package it's a package that's last been published 7 years ago. Do all the features still work as intented? I don't use NZBGet so I can't test.

jonjon1123 commented:
> Looks great! However my only concern is the usage of the `nzbget-api` package it's a package that's last been published 7 years ago. Do all the features still work as intented? I don't use NZBGet so I can't test.

Yeah, I had the same concern. From what I can tell, the nzbget api has not changed in a very long time. That _could_ explain the lack of updates. I have not tested all of api endpoints. Just the ones used for this use case.

JannesV commented:
> Yeah, I had the same concern. From what I can tell, the nzbget api has not changed in a very long time. That _could_ explain the lack of updates. I have not tested all of api endpoints. Just the ones used for this use case.

If all the functionality we currently leverage works that's good enough for me 😄

jonjon1123 commented on src/components/AppShelf/AddAppShelfItem.tsx:
```diff
@@ -407,6 +407,42 @@ export function AddAppShelfItemForm(props: AddAppShelfItemFormProps) {
+                    onChange={(event) => {
+                      form.setFieldValue('username', event.currentTarget.value);
+                    }}
+                    error={
```
To be honest, I just copied the code from the [Trasmission block](https://github.com/ajnart/homarr/blob/master/src/components/AppShelf/AddAppShelfItem.tsx#L374). It looks like your same question could be applied there too. So I guess the question now is, should I fix it? I could check to make sure `username` and `password` are not null/empty in the [form validation](https://github.com/ajnart/homarr/blob/master/src/components/AppShelf/AddAppShelfItem.tsx#L127).

manuel-rw replied:
I think that we shouldn't restrict the user. Are null / empty values invalid or impossible to use with NZBGet? If not, I think we shouldn't restrict it. Can we resolve this issue or is there still work left?

jonjon1123 replied:
I think I have taken care of this now. If the user enables the Usenet module, the newly added [NzbgetClient](https://github.com/jonjon1123/homarr/blob/282-nzbget-download-client/src/pages/api/modules/usenet/nzbget/nzbget-client.ts#L4) will check to make sure valid config is provided.

ajnart replied:
@jonjon1123 
sorry just found this after my comment. To check for null just add "required" to the fields classes. I removed it because some users had no user/pass and were complaining (although all logins worked)

ajnart replied:
Also ideally we'd move translations to something more "common" like password and username is not usenet specific. Perhaps it could be moved above in the translation tree

jonjon1123 replied:
Technically it isn't required to provide a username & password to use NZBGet with Homarr. It is only required if you are going to enable to Usenet module. That's why I added the check to the new [NzbgetClient](https://github.com/jonjon1123/homarr/blob/282-nzbget-download-client/src/pages/api/modules/usenet/nzbget/nzbget-client.ts#L4) that only gets invoked as part of the Usenet module.

manuel-rw reviewed (approved):
After this has been thoroughly tested with NZBget and all other issues have been resolved, we can merge this.
Don't forget to squash your commits into one and use https://gitmoji.dev/ for your commit message

manuel-rw commented:
@jonjon1123  Since this is a pretty major change, would you also volunteer to adjust the documentation, so users understand how to use it?
https://github.com/ajnart/homarr-docs

ajnart commented on package.json:
```diff
@@ -60,6 +60,7 @@
     "js-file-download": "^0.4.12",
     "next": "12.1.6",
     "next-i18next": "^11.3.0",
+    "nzbget-api": "0.0.3",
```
Is the indentation wrong ?

ajnart commented:
I'll do a more complete review but quick question. I saw you added some error handling with authetification like "Wrong password" (should be translated btw) 
but did you add the code to check that that these credentials do work ? If there is no validation function for that error handling it won't show up.

ajnart commented on src/tools/types.ts:
```diff
@@ -118,6 +120,7 @@ export const portmap = [
   { name: 'overseerr', value: '5055' },
   { name: 'dash.', value: '3001' },
   { name: 'sabnzbd', value: '8080' },
+  { name: 'nzbget', value: '6789' }
```
thank you for this

ajnart commented:
Thanks for all the hard work @jonjon1123 and nice work on actually creating types for everything and adding the ports to the portmap qol feature

manuel-rw commented:
> > @jonjon1123 Since this is a pretty major change, would you also volunteer to adjust the documentation, so users understand how to use it? https://github.com/ajnart/homarr-docs
> 
> Yeah, I will get this done.

Amazing! You made some really nice work here, thank you 🙏

ajnart commented on src/pages/api/modules/usenet/queue.ts:
```diff
@@ -33,42 +35,116 @@ async function Get(req: NextApiRequest, res: NextApiResponse) {
   }
 }
 
+function getNzbgetState(status: string) {
+  switch (status) {
+    case 'QUEUED':
+      return 'queued';
+    case 'PAUSED ':
+        return 'paused';
+    default:
+      return 'downloading';
+  }
+}
+
```
Some translation should be added here

jonjon1123 replied:
This is just a helper function to map the NZBget status to the `UsenetQueueItem` status options. From what I can tell, it isn't every displayed on screen.

ajnart replied:
oh so it's just an internal conversion, my bad

ajnart reviewed (commented):
I think some translations are missing. I haven't had time to review the actual integration yet. Every visible element on the page should be translated (It doesn't apply to error messages in your throws for example)

jonjon1123 commented:
> > > @jonjon1123 Since this is a pretty major change, would you also volunteer to adjust the documentation, so users understand how to use it? https://github.com/ajnart/homarr-docs
> > 
> > 
> > Yeah, I will get this done.
> 
> Amazing! You made some really nice work here, thank you 🙏

PR for the doc updates: https://github.com/ajnart/homarr-docs/pull/11

jonjon1123 commented:
Anything else I need to do to get this merged?

manuel-rw commented:
No, but I'd like to have a confirmation from someone else using NZBget that it's working.
I'll see whether I can quickly set it up locally after work and confirm that it works. If it does, we can merge

manuel-rw commented on src/pages/api/modules/usenet/nzbget/types.ts:
```diff
@@ -0,0 +1,135 @@
+  DownloadLimit: number;
+  ThreadCount: number;
+  PostJobCount: number;
+  UrlCount: number;
```
Do we really need stuff like this? I appreciate the type definition's, but it is kind of unnecessary, when we only use a few of them

JannesV replied:
This comment makes no sense lol 😄 This it the output from the NZBGet api, it's what's returned. You want to alter the types to not reflect reality just because it seems excessive?

manuel-rw replied:
No, you are right - We should keep them. I didn't think this quite through, sorry.

manuel-rw commented on src/pages/api/modules/usenet/pause.ts:
```diff
@@ -25,13 +26,43 @@ async function Post(req: NextApiRequest, res: NextApiResponse) {
 
-    const { origin } = new URL(service.url);
+        result = await new Promise((resolve, reject) => {
+          nzbGet.pauseDownload(false, (err: any, result: any) => {
```
Is there no definition available for ``err`` and ``result``?

jonjon1123 replied:
This is the callback that gets passed all the way to the underlying [`node-json-rpc`](https://www.npmjs.com/package/node-json-rpc) library. They don't provide any types for this callback in that library.

manuel-rw replied:
No worries. We'll close this then

manuel-rw commented on src/pages/api/modules/usenet/queue.ts:
```diff
@@ -33,42 +35,116 @@ async function Get(req: NextApiRequest, res: NextApiResponse) {
+          name: item.NZBName,
+          progress: (item.DownloadedSizeMB / item.FileSizeMB) * 100,
+          eta: (item.RemainingSizeMB * 1000000) / nzbgetStatus.DownloadRate,
+          // Multiple MB to get bytes
```
This comment is confusing. Why not ``Convert from bytes to MB``?

manuel-rw commented on src/pages/api/modules/usenet/resume.ts:
```diff
@@ -26,13 +27,43 @@ async function Post(req: NextApiRequest, res: NextApiResponse) {
 
-    const result = await new Client(origin, service.apiKey).queueResume();
+        result = await new Promise((resolve, reject) => {
+          nzbGet.resumeDownload(false, (err: any, result: any) => {
```
Same thing here. No type definitions?

jonjon1123 replied:
Same answer as above

manuel-rw reviewed (changes requested):
The config file is a no go! Please remove it from the history. Also, please squash your commit messages and reword to ``✨ Add NZBGet download client``

manuel-rw commented:
I was able to connect to a local docker instance of NZBGet. However, I wasn't able to add any newsservers in NZBGet - It told me that my credentials were wrong.
Not sure why that is happening though. But at least I can confirm, that it seems to be stable and shows any NZBs in the queue.
I think that we can merge after the issues I mentioned above have been resolved.

jonjon1123 commented:
> 

I believe I have now removed that commit from the history and squashed with the correct commit message.

jonjon1123 commented:
> The config file is a no go! Please remove it from the history. Also, please squash your commit messages and reword to `✨ Add NZBGet download client`

I believe I have now removed that commit from the history and squashed with the correct commit message.

ajnart commented:
> > The config file is a no go! Please remove it from the history. Also, please squash your commit messages and reword to `✨ Add NZBGet download client`
> 
> I believe I have now removed that commit from the history and squashed with the correct commit message.

it says 1000+ commits. Did you break it? 😅

jonjon1123 commented:
> > > The config file is a no go! Please remove it from the history. Also, please squash your commit messages and reword to `✨ Add NZBGet download client`
> > 
> > 
> > I believe I have now removed that commit from the history and squashed with the correct commit message.
> 
> it says 1000+ commits. Did you break it? 😅

Sorry, struggling to completely remove the secret. Thought I had it. Guess not. :(

manuel-rw commented:
``Merge branch 'master' of github.com:jonjon1123/homarr`` -> looks like you rebased to master instead of ``dev``

manuel-rw commented:
Config still seems to be in the history. @ajnart can you maybe fix the rebase?

ajnart commented:
Not sure how to fix it

jonjon1123 commented:
I am not sure what happened, but it doesn't seem to be easily fixable. I created a new PR:https://github.com/ajnart/homarr/pull/497. Can we use this new one instead and I can delete this old one?
